### DISCUSSION TITLE
Are things like ::webkit-scrollbar in scope?

### DISCUSSION THREAD

**Discussion by cly** [css-hooks/css-hooks#70] [General]
I was trying to create a hook for webkit-scrollbar and was not able to make it work.

Tried different combinations of `&::-webkit-scrollbar` etc.

What I would like to do is hide the scrollbar via appearance.

When I used the `&::-webkit-scrollbar` hook, the entire page lost its scrollbar.

Any ideas?

Thanks!

**Comment by nsaunders** [comment]
Hi @cly, CSS Hooks is primarily focused on conditional styling. Many CSS rules fit this category, including:
- pseudo-classes (styles apply on a given state, e.g. `:hover`)
- class/id selectors (styles apply when a given class name or id exists on the element)
- media/container queries (styles apply based on element/viewport dimensions)
- feature queries (styles apply when the browser supports a particular feature)

Pseudo-elements such as `::before`, `::placeholder`, `::-webkit-scrollbar`, etc., don't fit that profile unfortunately.

If you haven't already, may I suggest having a look at the [Introduction](https://css-hooks.com/docs/introduction) which explains how CSS Hooks works "under the hood". I think this additional context might help with understanding why pseudo-elements are out of scope.

Here's how I typically work around this limitation:
1. For `::before`, `::after`, or even `::placeholder`, I will simply add a real DOM element instead of using a pseudo-element, and apply the required styling to that element instead. `::before` is roughly equivalent to an extra element inserted as the first child, while `::after` is like an extra element inserted as the last child. I have a similar approach for `::placeholder`, but I also use the `:placeholder-shown` pseudo-class (supported by hooks) to control the visibility of the placeholder element.
2. Maintaining a global style sheet the "traditional" way. The CSS Hooks approach, as powerful as it is, doesn't have an answer to every CSS feature (`@keyframes`, `@font-face`, and of course pseudo-elements...etc.) Indeed, CSS is still required, but hooks help me to keep the amount small and maintainable.

Hope this helps, and please let me know if you still have questions.

**Comment by cly** [comment]
Thanks, just wanted to confirm! I wonder if a future version could take inspiration from stylexjs and support these use cases.

https://stylexjs.com/docs/api/javascript/keyframes/
https://stylexjs.com/docs/learn/styling-ui/defining-styles/#:~:text=Pseudo%2Delements%E2%80%8B&text=This%20helps%20reduce%20the%20size,an%20input%20or%20textarea%20element.

fwiw, stylexjs does not support the rich selectors of css hooks which causes problems when working with third party libs and decedent selectors etc.

**Comment by robinweser** [comment]
@cly I have support keyframes in [brandeur](https://github.com/robinweser/brandeur) (built on top of css-hooks) that is very similar to the stylex one. Maybe eyou want to check it out 🙃

**Reply by cly** [reply]
Cool! Thanks!

**Comment by cly** [comment]
I have somewhat of a workaround here. Might be helpful for others.

Just to recap, I'd like to hide scrollbar on an overflown element which requires targeting ::webkit-scrollbar.

To a global style sheet I added.
```
::-webkit-scrollbar {
    display: var(--webkit-scrollbar-display, initial);
}
```

And then via css hooks I can just simply set the variable.
```
css({
  "--webkit-scrollbar-display": "none"
})
 ```
 
 A small style sheet which makes all pseudo elements targetable could be done with something like this I think?
 
 Caveat: children will also inherit this and will need to unset it if needed.

**Comment by nsaunders** [comment]
Thanks for sharing you workaround @cly. That would be my approach as well.

I'm going to go ahead and close the discussion for now, but please let me know if any additional thoughts come to mind regarding pseudo-elements.
